File: core/remote/frames.mjs
// リモート接続の平文のフレーム（docs/remote.md §4.1）。
//
//   type (u8) | stream (u32, BE) | payload
//
// 1 フレーム = Noise の 1 通 = 中継への WebSocket の 1 メッセージ（バイナリ）。
// stream 0 はチャネル自体。端末が開くストリームは奇数、ホストが開くもの（今は無い）は偶数。
// ここはバイト列との変換だけ。流量の制御・ストリームの状態は channel.mjs が持つ。
import { MAX_PLAINTEXT } from './noise.mjs';

export const T = Object.freeze({
  HELLO: 0x01,
  PING: 0x02,
  PONG: 0x03,
  GOAWAY: 0x04,
  HTTP_REQ: 0x10,
  HTTP_RES: 0x11,
  DATA: 0x12,
  END: 0x13,
  RESET: 0x14,
  WS_OPEN: 0x20,
  WS_ACCEPT: 0x21,
  WS_REJECT: 0x22,
  WS_MSG: 0x23,
  WS_CLOSE: 0x24,
  WINDOW: 0x30,
});

export const TYPE_NAMES = Object.freeze(Object.fromEntries(Object.entries(T).map(([k, v]) => [v, k])));

/** stream 0 だけに載る型と、stream 0 には載らない型。 */
const CHANNEL_ONLY = new Set([T.HELLO, T.PING, T.PONG, T.GOAWAY]);

export const HEADER_BYTES = 5;
/** 平文のフレームの上限（Noise の 1 通 65535 - タグ 16）。 */
export const MAX_FRAME = MAX_PLAINTEXT;
export const MAX_PAYLOAD = MAX_FRAME - HEADER_BYTES;
/** 本文・WebSocket のメッセージはこの大きさごとに分ける。 */
export const CHUNK = 60 * 1024;
export const PROTO = 1;

/** WS_MSG の印（payload の先頭 1 バイト）。 */
export const WS_TEXT = 0x01;
export const WS_FIN = 0x02;

/** RESET の理由（u16）。docs/remote.md §4.1 に表を置いた。 */
export const RESET_CODE = Object.freeze({
  CANCEL: 0,          // 片方が要らなくなった（窓を閉じた・要求を取り消した）
  PROTOCOL: 1,        // 決まりに反するフレーム
  REFUSED: 2,         // 同時ストリームの上限・受け付けない種類
  FORBIDDEN: 3,       // 接続口の防火壁が通さない（/mcp/・GET/HEAD 以外・/ws 以外の WS）
  INTERNAL: 4,        // 受け側の失敗（ローカルのサーバーにつながらないなど）
  FLOW_CONTROL: 5,    // 窓を超えて送られた
  CHANNEL_CLOSED: 6,  // チャネルが閉じた（ローカルで付く。送られはしない）
  TOO_LARGE: 7,       // 組み立てた WebSocket のメッセージが上限を超えた
});

export class FrameError extends Error {}

function bytes(v) {
  if (v == null) return Buffer.alloc(0);
  if (Buffer.isBuffer(v)) return v;
  if (v instanceof Uint8Array) return Buffer.from(v.buffer, v.byteOffset, v.byteLength);
  if (typeof v === 'string') return Buffer.from(v, 'utf8');
  throw new TypeError('payload はバイト列か文字列');
}

export function encodeFrame(type, stream, payload) {
  if (!TYPE_NAMES[type]) throw new FrameError(`知らない型: ${type}`);
  if (!Number.isInteger(stream) || stream < 0 || stream > 0xffffffff) throw new FrameError(`stream が範囲外: ${stream}`);
  const body = bytes(payload);
  if (body.length > MAX_PAYLOAD) throw new FrameError(`payload が大きすぎる（${body.length} > ${MAX_PAYLOAD}）`);
  const out = Buffer.allocUnsafe(HEADER_BYTES + body.length);
  out.writeUInt8(type, 0);
  out.writeUInt32BE(stream, 1);
  body.copy(out, HEADER_BYTES);
  return out;
}

/** バイト列 → `{ type, stream, payload }`。形の誤り（短い・知らない型・stream の取り違え）は FrameError。 */
export function decodeFrame(buf) {
  buf = bytes(buf);
  if (buf.length < HEADER_BYTES) throw new FrameError('フレームが短すぎる');
  if (buf.length > MAX_FRAME) throw new FrameError('フレームが大きすぎる');
  const type = buf.readUInt8(0);
  const stream = buf.readUInt32BE(1);
  if (!TYPE_NAMES[type]) throw new FrameError(`知らない型: 0x${type.toString(16)}`);
  if (CHANNEL_ONLY.has(type) && stream !== 0) throw new FrameError(`${TYPE_NAMES[type]} は stream 0 だけ`);
  if (!CHANNEL_ONLY.has(type) && type !== T.WINDOW && stream === 0) throw new FrameError(`${TYPE_NAMES[type]} は stream 0 に載らない`);
  return { type, stream, payload: buf.subarray(HEADER_BYTES) };
}

// ── payload の形 ───────────────────────────────────────────

export const json = {
  encode: v => Buffer.from(JSON.stringify(v), 'utf8'),
  decode(buf) {
    let v;
    try { v = JSON.parse(bytes(buf).toString('utf8')); } catch { throw new FrameError('JSON として読めない'); }
    if (!v || typeof v !== 'object' || Array.isArray(v)) throw new FrameError('JSON のオブジェクトではない');
    return v;
  },
};

export function u16(n) { const b = Buffer.alloc(2); b.writeUInt16BE(n); return b; }
export function readU16(buf) {
  if (buf.length !== 2) throw new FrameError('u16 の長さが違う');
  return buf.readUInt16BE(0);
}
export function u32(n) { const b = Buffer.alloc(4); b.writeUInt32BE(n); return b; }
export function readU32(buf) {
  if (buf.length !== 4) throw new FrameError('u32 の長さが違う');
  return buf.readUInt32BE(0);
}

/** WS_CLOSE: u16 の close code + 理由（UTF-8。WebSocket と同じく 123 バイトまでに切る）。 */
export function encodeWsClose(code = 1000, reason = '') {
  let r = Buffer.from(String(reason ?? ''), 'utf8');
  if (r.length > 123) r = r.subarray(0, 123);
  return Buffer.concat([u16(code), r]);
}
export function decodeWsClose(buf) {
  if (buf.length < 2) throw new FrameError('WS_CLOSE が短すぎる');
  return { code: buf.readUInt16BE(0), reason: buf.subarray(2).toString('utf8') };
}

/** WS_MSG の 1 断片: 印（bit0 文字、bit1 最後の断片）+ 断片。 */
export function encodeWsFragment(chunk, { text = false, fin = true } = {}) {
  return Buffer.concat([Buffer.from([(text ? WS_TEXT : 0) | (fin ? WS_FIN : 0)]), bytes(chunk)]);
}
export function decodeWsFragment(buf) {
  if (buf.length < 1) throw new FrameError('WS_MSG が短すぎる');
  const flags = buf.readUInt8(0);
  if (flags & ~(WS_TEXT | WS_FIN)) throw new FrameError('WS_MSG の印に知らないビット');
  return { text: Boolean(flags & WS_TEXT), fin: Boolean(flags & WS_FIN), data: buf.subarray(1) };
}

/** 大きいバイト列を CHUNK ごとに分ける。空なら空の配列。 */
export function chunks(buf, size = CHUNK) {
  buf = bytes(buf);
  const out = [];
  for (let off = 0; off < buf.length; off += size) out.push(buf.subarray(off, off + size));
  return out;
}

/** WebSocket の 1 メッセージを WS_MSG の payload の列にする（最後の断片にだけ fin）。空のメッセージも 1 断片。 */
export function fragmentWsMessage(data, { text = typeof data === 'string', size = CHUNK } = {}) {
  const parts = chunks(data, size);
  if (!parts.length) parts.push(Buffer.alloc(0));
  return parts.map((p, i) => encodeWsFragment(p, { text, fin: i === parts.length - 1 }));
}

/**
 * WS_MSG の断片を組み立てる。最初の断片の印で文字 / バイナリを決め、途中で変わったら誤り。
 * 組み立て途中の大きさが maxBytes を超えたら FrameError（RESET TOO_LARGE にする）。
 */
export class WsAssembler {
  constructor({ maxBytes = 64 * 1024 * 1024 } = {}) {
    this.maxBytes = maxBytes;
    this.parts = [];
    this.size = 0;
    this.text = null;
  }
  get pending() { return this.text != null; }
  /** 断片を足す。メッセージが揃ったら `{ data, text }`、まだなら null。 */
  push(payload) {
    const f = decodeWsFragment(bytes(payload));
    if (this.text == null) this.text = f.text;
    else if (this.text !== f.text) throw new FrameError('WS_MSG の文字 / バイナリが途中で変わった');
    this.size += f.data.length;
    if (this.size > this.maxBytes) throw new FrameError('WebSocket のメッセージが上限を超えた');
    this.parts.push(f.data);
    if (!f.fin) return null;
    const data = this.parts.length === 1 ? Buffer.from(this.parts[0]) : Buffer.concat(this.parts);
    const text = this.text;
    this.parts = [];
    this.size = 0;
    this.text = null;
    return { data, text };
  }
}

File: mobile/android/remote-core/src/main/kotlin/com/procway/pleiad/remote/Frames.kt
package com.procway.pleiad.remote

// Kotlin port of core/remote/frames.mjs (docs/remote.md §4.1).
//   type (u8) | stream (u32, BE) | payload
// One frame = one Noise message = one binary WebSocket message to the relay.

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import org.json.JSONObject

object T {
    const val HELLO = 0x01
    const val PING = 0x02
    const val PONG = 0x03
    const val GOAWAY = 0x04
    const val HTTP_REQ = 0x10
    const val HTTP_RES = 0x11
    const val DATA = 0x12
    const val END = 0x13
    const val RESET = 0x14
    const val WS_OPEN = 0x20
    const val WS_ACCEPT = 0x21
    const val WS_REJECT = 0x22
    const val WS_MSG = 0x23
    const val WS_CLOSE = 0x24
    const val WINDOW = 0x30

    val NAMES = mapOf(
        HELLO to "HELLO", PING to "PING", PONG to "PONG", GOAWAY to "GOAWAY",
        HTTP_REQ to "HTTP_REQ", HTTP_RES to "HTTP_RES", DATA to "DATA", END to "END", RESET to "RESET",
        WS_OPEN to "WS_OPEN", WS_ACCEPT to "WS_ACCEPT", WS_REJECT to "WS_REJECT", WS_MSG to "WS_MSG", WS_CLOSE to "WS_CLOSE",
        WINDOW to "WINDOW",
    )
    internal val CHANNEL_ONLY = setOf(HELLO, PING, PONG, GOAWAY)
}

object ResetCode {
    const val CANCEL = 0
    const val PROTOCOL = 1
    const val REFUSED = 2
    const val FORBIDDEN = 3
    const val INTERNAL = 4
    const val FLOW_CONTROL = 5
    const val CHANNEL_CLOSED = 6
    const val TOO_LARGE = 7
}

class FrameError(msg: String, val tooLarge: Boolean = false) : Exception(msg)

class Frame(val type: Int, val stream: Long, val payload: ByteArray)

object Frames {
    const val HEADER_BYTES = 5
    const val MAX_FRAME = NoiseConst.MAX_PLAINTEXT
    const val MAX_PAYLOAD = MAX_FRAME - HEADER_BYTES
    const val CHUNK = 60 * 1024
    const val PROTO = 1
    const val WS_TEXT = 0x01
    const val WS_FIN = 0x02

    fun encode(type: Int, stream: Long, payload: ByteArray?): ByteArray {
        if (type !in T.NAMES) throw FrameError("unknown type: $type")
        if (stream < 0 || stream > 0xffffffffL) throw FrameError("stream out of range: $stream")
        val body = payload ?: ByteArray(0)
        if (body.size > MAX_PAYLOAD) throw FrameError("payload too large (${body.size} > $MAX_PAYLOAD)")
        val out = ByteBuffer.allocate(HEADER_BYTES + body.size)
        out.put(type.toByte())
        out.putInt(stream.toInt())
        out.put(body)
        return out.array()
    }

    fun decode(buf: ByteArray): Frame {
        if (buf.size < HEADER_BYTES) throw FrameError("frame too short")
        if (buf.size > MAX_FRAME) throw FrameError("frame too large")
        val type = buf[0].toInt() and 0xff
        val stream = ByteBuffer.wrap(buf, 1, 4).int.toLong() and 0xffffffffL
        val name = T.NAMES[type] ?: throw FrameError("unknown type: 0x${type.toString(16)}")
        if (type in T.CHANNEL_ONLY && stream != 0L) throw FrameError("$name is only allowed on stream 0")
        if (type !in T.CHANNEL_ONLY && type != T.WINDOW && stream == 0L) throw FrameError("$name is not allowed on stream 0")
        return Frame(type, stream, buf.copyOfRange(HEADER_BYTES, buf.size))
    }

    fun jsonEncode(v: JSONObject): ByteArray = v.toString().toByteArray(Charsets.UTF_8)

    fun jsonDecode(buf: ByteArray): JSONObject = try {
        JSONObject(String(buf, Charsets.UTF_8))
    } catch (e: Exception) {
        throw FrameError("invalid JSON")
    }

    fun u16(n: Int): ByteArray = byteArrayOf((n ushr 8).toByte(), n.toByte())
    fun readU16(b: ByteArray): Int {
        if (b.size != 2) throw FrameError("bad u16 length")
        return ((b[0].toInt() and 0xff) shl 8) or (b[1].toInt() and 0xff)
    }
    fun u32(n: Long): ByteArray = ByteBuffer.allocate(4).putInt(n.toInt()).array()
    fun readU32(b: ByteArray): Long {
        if (b.size != 4) throw FrameError("bad u32 length")
        return ByteBuffer.wrap(b).int.toLong() and 0xffffffffL
    }

    fun encodeWsClose(code: Int = 1000, reason: String = ""): ByteArray {
        var r = reason.toByteArray(Charsets.UTF_8)
        if (r.size > 123) r = r.copyOf(123)
        return u16(code) + r
    }

    fun decodeWsClose(b: ByteArray): Pair<Int, String> {
        if (b.size < 2) throw FrameError("WS_CLOSE too short")
        return readU16(b.copyOf(2)) to String(b, 2, b.size - 2, Charsets.UTF_8)
    }

    fun encodeWsFragment(chunk: ByteArray, text: Boolean, fin: Boolean): ByteArray =
        byteArrayOf(((if (text) WS_TEXT else 0) or (if (fin) WS_FIN else 0)).toByte()) + chunk

    class WsFragment(val text: Boolean, val fin: Boolean, val data: ByteArray)

    fun decodeWsFragment(b: ByteArray): WsFragment {
        if (b.isEmpty()) throw FrameError("WS_MSG too short")
        val flags = b[0].toInt() and 0xff
        if (flags and (WS_TEXT or WS_FIN).inv() != 0) throw FrameError("unknown WS_MSG flag bits")
        return WsFragment(flags and WS_TEXT != 0, flags and WS_FIN != 0, b.copyOfRange(1, b.size))
    }
}

/** Reassembles WS_MSG fragments. Exceeding maxBytes throws FrameError(tooLarge = true) (RESET TOO_LARGE). */
class WsAssembler(private val maxBytes: Long = 64L * 1024 * 1024) {
    private var parts = ByteArrayOutputStream()
    private var text: Boolean? = null

    class Message(val data: ByteArray, val text: Boolean)

    fun push(payload: ByteArray): Message? {
        val f = Frames.decodeWsFragment(payload)
        val t = text
        if (t == null) text = f.text
        else if (t != f.text) throw FrameError("WS_MSG switched between text and binary mid-message")
        if (parts.size().toLong() + f.data.size > maxBytes) throw FrameError("WebSocket message exceeds the limit", tooLarge = true)
        parts.write(f.data)
        if (!f.fin) return null
        val msg = Message(parts.toByteArray(), text!!)
        parts = ByteArrayOutputStream()
        text = null
        return msg
    }
}
